### PR TITLE
fix problem with mathml namespaces

### PR DESCRIPTION
**Pull request by wilkus27**
Issue: https://github.com/openstax/cnx-recipes/issues/5142

### CODE REVIEW

**Comment by philschatz** [comment]
I'm not very familiar with Ruby but would a method like this on `document.rb` followed by a call to it in `oven.py` work?

```ruby
    # document.rb
    def ensure_namespaces
      traverse do |node|
        next unless node.element?

        node.attribute_nodes.each do |attr|
          next unless attr.namespace

          unless node.has_namespace_defined_on_ancestor?(attribute: attr)
            node.add_namespace(attr.namespace.prefix, attr.namespace.href)
            # Might also need to set the `xmlns:{...}="..."` attribute
          end
        end

        next unless node.namespace

        unless node.has_namespace_defined_on_ancestor?
          node.add_namespace(node.namespace.prefix, node.namespace.href)
          # Might also need to set the `xmlns:{...}="..."` attribute
        end
      end
    end
```

and the corresponding one-line change in oven.rb :+1: 

```ruby
      doc.ensure_namespaces
      File.open(output_file, 'w') do |f|
        ...
```

**Comment by wilkus27** [comment]
@philschatz It doesn't work because this part is removing namespaces:

image

I checked it with this:

image

If element was not moved it still has a namespaces:

image

But one that was moved during baking (with `cut`/`clone` and `paste` methods), doesn't have one:

image

So the rest of the code doesn't affect it:

image

Calling `ensure_namespaces` before part with baking also doesn't work, but i don't know why :thinking: 

image

Namespaces are present after calling it:

image

After baking they're not:

image

image

File: lib/kitchen/element_base.rb [hunk 1]
```diff
@@ -918,7 +935,7 @@ def rex_link
       page_string = ''
       page_title = ''
       page = element_with_ancestors.ancestor(:page) if element_with_ancestors.has_ancestor?(:page)
-      if page && page&.is_introduction?
+      if page&.is_introduction?
```
Comment (wilkus27) [inline lib/kitchen/element_base.rb:938]: Rubocop was complaining about that after updating to Ruby 3

File: lib/kitchen/patches/nokogiri.rb [hunk 2]
```diff
@@ -69,12 +69,28 @@ def classes
 
       def preceded_by_text?
         prev = previous_sibling
-        while !prev.nil? && prev.blank? do prev = prev.previous_sibling end
+        prev = prev.previous_sibling while !prev.nil? && prev.blank?
```
Comment (wilkus27) [inline lib/kitchen/patches/nokogiri.rb:72]: Rubocop was complaining about that after updating to Ruby 3

File: lib/kitchen/patches/nokogiri.rb [hunk 3]
```diff
@@ -83,7 +99,6 @@ def self.selector_to_css_nodes(selector)
 
       protected
 
-      # rubocop:disable Metrics/CyclomaticComplexity
```
Comment (wilkus27) [inline lib/kitchen/patches/nokogiri.rb:86]: Rubocop was complaining about that after updating to Ruby 3

File: lib/kitchen/element_base.rb [hunk 4]
```diff
@@ -800,6 +800,32 @@ def to_xhtml
...
+        # Define all namespaces that are used inside the node being cloned.
+        #
+        # When document is created, all namespaces definitions from the book
+        # are collected and added to the html tag.
+        # In the moment when clipboard is pasted and converted to string
+        # we are loosing information about namespaces defined on html ancestor.
+        #
+        # It's important to add namespace definition directly to the node that has a namespace.
+        # Adding it just to the node being cloned is breaking elements that have
+        # namespaces without prefix (e.g <math xmlns="http://www.w3.org/1998/Math/MathML"/>)
```
Comment (kswanson33) [inline lib/kitchen/element_base.rb:812]: 👏

**Review by kswanson33 — APPROVED** [review]
this clear documentation though 💪 tysm natalia!

**Pull request merged**